workflow/controller/operator.go

```diff
@@ -1871,13 +1884,15 @@ func (woc *wfOperationCtx) executeTemplate(ctx context.Context, nodeName string,
 			if node == nil {
 				node = woc.initializeExecutableNode(nodeName, wfutil.GetNodeType(processedTmpl), templateScope, processedTmpl, orgTmpl, opts.boundaryID, wfv1.NodePending, msg)
 			}
-			lockName, err := argosync.GetLockName(processedTmpl.Synchronization, woc.wf.Namespace)
-			if err != nil {
-				// If an error were to be returned here, it would have been caught by TryAcquire. If it didn't, then it is
-				// unexpected behavior and is a bug.
-				panic("bug: GetLockName should not return an error after a call to TryAcquire")
+			if !woc.releaseLocksForPendingShuttingdownWfs(ctx) {
+				lockName, err := argosync.GetLockName(processedTmpl.Synchronization, woc.wf.Namespace)
+				if err != nil {
+					// If an error were to be returned here, it would have been caught by TryAcquire. If it didn't, then it is
+					// unexpected behavior and is a bug.
+					panic("bug: GetLockName should not return an error after a call to TryAcquire")
+				}
+				return woc.markNodeWaitingForLock(node.Name, lockName.EncodeName()), nil
```

Comment:
Sorry, I just noticed this.

I don't think it is correct to be releasing locks as part of `executeTemplate()` when terminating a Running workflow. Because at this point, the workflow is already Running and has acquired the lock.  By releasing the locks here, I believe we will violate mutex and semaphore. Why not let it release naturally?

Reply:
Good catch indeed. I just removed this part of the logic.